File: profiler.py
#!/usr/bin/env python3
"""
profiler.py -- Parse Gradle build logs and show slowest tasks
Usage: ./gradlew build 2>&1 | python3 profiler.py
       Or: python3 profiler.py build.log
"""
import sys, re
from collections import defaultdict
from datetime import timedelta

def parse_gradle_log(lines):
    tasks = []
    for line in lines:
        # Pattern: > Task :app:compileDebugJavaWithJavac FAILED in 2s 345ms
        m = re.search(r'> Task (:[\w:]+)\s+(\w+)\s+in\s+([\d.]+)(ms|s)', line)
        if m:
            task_name = m.group(1)
            status = m.group(2)
            duration = float(m.group(3))
            unit = m.group(4)
            if unit == 's':
                duration *= 1000
            tasks.append({'name': task_name, 'duration': duration, 'status': status})
    return tasks

def format_duration(ms):
    if ms < 1000:
        return f"{ms:.0f}ms"
    return f"{ms/1000:.2f}s"

def main():
    if len(sys.argv) > 1:
        with open(sys.argv[1]) as f:
            lines = f.readlines()
    else:
        lines = sys.stdin.readlines()

    tasks = parse_gradle_log(lines)
    if not tasks:
        print("No Gradle tasks found in log.")
        return

    tasks.sort(key=lambda t: t['duration'], reverse=True)
    total_time = sum(t['duration'] for t in tasks)

    print(f"\n⚙️  Gradle Build Profile\n")
    print(f"{'Task':<45} {'Time':<10} {'% of Total'}")
    print("─" * 65)
    for t in tasks[:20]:
        pct = (t['duration'] / total_time * 100) if total_time else 0
        print(f"{t['name']:<45} {format_duration(t['duration']):<10} {pct:>5.1f}%")

    print(f"\n{'Total build time:':<45} {format_duration(total_time)}")
    print(f"{'Tasks profiled:':<45} {len(tasks)}")

if __name__ == "__main__":
    main()

File: Profiler.kt
package com.outrageousstorm.gradle

import java.io.File
import kotlin.math.roundToInt

data class GradleTask(
    val name: String,
    val duration: Long,  // ms
    val plugin: String,
    val deps: List<String> = emptyList()
)

class GradleProfiler(val profileHtml: String) {
    fun parse(): List<GradleTask> {
        val tasks = mutableListOf<GradleTask>()
        val content = File(profileHtml).readText()
        
        // Parse HTML table with tasks
        // Format: each row is <tr><td>taskname</td><td>duration</td></tr>
        val taskRegex = """<tr><td>([^<]+)</td><td>([\d.]+)s</td><td>([^<]*)</td></tr>""".toRegex()
        taskRegex.findAll(content).forEach { match ->
            val (name, duration, plugin) = match.destructured
            tasks.add(GradleTask(
                name = name.trim(),
                duration = (duration.toFloat() * 1000).toLong(),
                plugin = plugin.trim()
            ))
        }
        
        return tasks.sortedByDescending { it.duration }
    }
    
    fun criticalPath(): List<GradleTask> {
        val tasks = parse()
        return tasks.take((tasks.size * 0.2).roundToInt().coerceAtLeast(5))
    }
    
    fun report(): String {
        val tasks = parse()
        val total = tasks.sumOf { it.duration }
        
        return buildString {
            appendLine("=== Gradle Build Profile ===\n")
            appendLine("Total time: ${total / 1000}s\n")
            
            appendLine("Top 10 slowest tasks:")
            tasks.take(10).forEachIndexed { i, task ->
                val pct = (task.duration * 100 / total).toInt()
                appendLine("  ${i+1}. ${task.name.padEnd(40)} ${task.duration}ms ($pct%)")
            }
            
            appendLine("\nCritical path (blocking tasks):")
            criticalPath().forEach { task ->
                appendLine("  • ${task.name} (${task.duration}ms)")
            }
            
            appendLine("\nOptimization tips:")
            appendLine("  1. Use KSP instead of Kapt if using annotation processors")
            appendLine("  2. Enable --parallel in gradle.properties")
            appendLine("  3. Set org.gradle.parallel=true")
            appendLine("  4. Increase org.gradle.workers.max=8 (or CPU count)")
            appendLine("  5. Enable build cache: org.gradle.caching=true")
        }
    }
}

fun main(args: Array<String>) {
    if (args.isEmpty()) {
        println("Usage: java -jar profiler.jar <build-profile.html>")
        return
    }
    
    val profiler = GradleProfiler(args[0])
    println(profiler.report())
}
